File: Public/Gui/Linn/Gui/SceneGraph/Node.cs
using Linn;
using Linn.Gui;
using Linn.Gui.Resources;
using Linn.Gui.Scenegraph;
using System.Collections;
using System.Collections.Generic;
using System.Xml;
using System;
using System.Threading;

namespace Linn {
namespace Gui {
namespace Scenegraph {

public class DuplicateNode : System.Exception {
    public DuplicateNode(string aNode) : base(aNode) {}
}

internal class InvalidParent : System.Exception
{
}

public class Node : Plugin, IMessengerObserver, ICloneable
{
    public class NearToFar : IComparer<Node>
    {
        public int Compare(Node x, Node y) {
            return(x.WorldSrt.Translation.Z.CompareTo(y.WorldSrt.Translation.Z));
        }
    }
    
    public class FarToNear : IComparer<Node>
    {
        public int Compare(Node x, Node y) {
            return(y.WorldSrt.Translation.Z.CompareTo(x.WorldSrt.Translation.Z));
        }
    }

    public Node() : base(UniqueName("Node")) {
    }
    
    public Node(string aName) : base(aName) {
    }
    
    public override Object Clone() {
        Node n = new Node();        
        Clone(n);
        return n;
    }
    
    protected void Clone(Node aNode) {
        Clone((Plugin)aNode);
        aNode.iRoot = iRoot;
        aNode.iActive = iActive;
        aNode.iParent = iParent;
        aNode.iChildrenList.AddRange(iChildrenList);
        aNode.iCompositeChildrenList.AddRange(iCompositeChildrenList);
        aNode.iTranslatorIn = iTranslatorIn;    // shallow copy
        aNode.iTranslatorOut = iTranslatorOut;  // shallow copy
        aNode.iLocalSrt = (Srt)iLocalSrt.Clone();
        aNode.iWorldSrt = (Srt)iWorldSrt.Clone();
    }
    
    public override void Load(XmlNode aXmlNode) {
        base.Load(aXmlNode);
        
        XmlNodeList list;
        list = aXmlNode.SelectNodes("Parent");
        if(list != null) {
            Assert.Check(list.Count == 1);
            if(list[0].FirstChild != null) {
                iParent = new ReferencePlugin<Node>(list[0].FirstChild.Value);
                /*if(list[0].Value == "") {
                    iRoot = true;
                }*/
            } else {
                iRoot = true;
            }
        }
        
        list = aXmlNode.SelectNodes("Active");
        if(list != null) {
            Assert.Check(list.Count == 1);
            iActive = bool.Parse(list[0].FirstChild.Value);
        }
        
        list = aXmlNode.SelectNodes("Srt");
        if(list != null) {
            Assert.Check(list.Count == 1);
            iLocalSrt.Load(list[0]);
        }
        
        list = aXmlNode.SelectNodes("InMessageTranslator");
        if(list != null) {
            Assert.Check(list.Count == 1);
            iTranslatorIn.Load(list[0]);
        }

        list = aXmlNode.SelectNodes("OutMessageTranslator");
        if(list != null) {
            Assert.Check(list.Count == 1);
            iTranslatorOut.Load(list[0]);
        }
        
        list = aXmlNode.SelectNodes("Children");
        if(list != null) {
            Assert.Check(list.Count == 1);
            list = list[0].SelectNodes("Child");
            foreach(XmlNode node in list) {
                ReferencePlugin<Node> child = new ReferencePlugin<Node>(node.FirstChild.Value);
                iChildrenList.Add(child);
                // TODO: check for duplicate children
            }
        }
    }
    
    public override void Link() {
        //Trace.WriteLine(Trace.kGui, "Linking Node: " + Fullname);
        base.Link();
        bool refNode = (Parent != null);
        iParent.Link();
        // we need to add nodes that are child nodes in external packages
        if(Parent != null && !refNode) {
            if(Parent.Namespace != Namespace) {
                Node node = Parent;
                Parent = null;
                node.AddChild(this);
            }
        }
        iTranslatorIn.Link();
        iTranslatorOut.Link();
        foreach(ReferencePlugin<Node> child in iChildrenList) {
            child.Link();
            // when need to set the correct parent for root nodes that
            // are referenced in external packages
            if(child.Object.iParent.Name == "") {
                child.Object.Parent = this;
            }
        }
        Update(false, true);
    }
    
    public override void Save(XmlTextWriter aWriter) {
        base.Save(aWriter);
        aWriter.WriteStartElement("Parent");
        if(Parent != null && !iRoot) {
            aWriter.WriteString(Parent.Fullname);
        }
        aWriter.WriteEndElement();
        aWriter.WriteStartElement("Active");
        aWriter.WriteString(iActive.ToString());
        aWriter.WriteEndElement();
        
        iLocalSrt.Save(aWriter);
        
        aWriter.WriteStartElement("InMessageTranslator");
        iTranslatorIn.Save(aWriter);
        aWriter.WriteEndElement();
        
        aWriter.WriteStartElement("OutMessageTranslator");
        iTranslatorOut.Save(aWriter);
        aWriter.WriteEndElement();
        
        aWriter.WriteStartElement("Children");
        foreach(ReferencePlugin<Node> node in iChildrenList) {
            aWriter.WriteStartElement("Child");
            aWriter.WriteString(node.Object.Fullname);
            aWriter.WriteEndElement();
        }   
        aWriter.WriteEndElement();
    }
    
    public bool Active {
        get {
            return iActive;
        }
        set {
            bool update = false;
            Lock();
            if(iActive != value) {
                iActive = value;
                Update(false, false);
                foreach(ReferencePlugin<Node> child in iChildrenList) {
                    child.Object.Active = value;
                }
                foreach(ReferencePlugin<Node> child in iCompositeChildrenList) {
                    child.Object.Active = value;
                }
                update = true;
            }
            Unlock();
            if(update) {
                OnSetActive();
            }
        }
    }
    
    public virtual void OnSetActive() {
        MsgActiveChanged msg = new MsgActiveChanged(this, !iActive, iActive);
        SendMessage(msg);
    }
    
    public Node Parent {
        get {
            return iParent.Object;
        }
        set {
            Lock();
            iParent = new ReferencePlugin<Node>(value);
            Unlock();
            ObserverUpdate();
        }
    }
    
    public bool IsRoot {
        get {
            return iRoot;
        }
    }
    
    public Srt LocalSrt {
        get {
            return iLocalSrt;
        }
    }
    
    public Srt WorldSrt {
        get {
            return iWorldSrt;
        }
    }
    
    public virtual Vector3d LocalTranslation {
        get {
            return iLocalSrt.Translation;
        }
        set {
            Lock();
            iLocalSrt.Translation = value;
            Update(true, false);
            Unlock();
        }
    }
    
    public virtual Vector3d WorldTranslation {
        get {
            return iWorldSrt.Translation;
        }
        set {
            Lock();
            iWorldSrt.Translation = value;
            //Trace.WriteLine(Trace.kGui, "Node.WorldTranslation: parent = " + Parent);
            if(Parent != null) {
                iLocalSrt.Translation = iWorldSrt.Translation - Parent.WorldTranslation;
                //Trace.WriteLine(Trace.kGui, Fullname + ": WorldTrans = " + iWorldSrt.Translation);
                //Trace.WriteLine(Trace.kGui, Fullname + ": Parent WorldTrans = " + Parent.WorldTranslation);
                //Trace.WriteLine(Trace.kGui, Fullname + ": LocalTrans = " + iLocalSrt.Translation);
            }
            else {
                iLocalSrt.Translation = value;
                //Trace.WriteLine(Trace.kGui, Fullname + ": WorldTrans = " + iWorldSrt.Translation);
                //Trace.WriteLine(Trace.kGui, Fullname + ": LocalTrans = " + iLocalSrt.Translation);
            }
            Update(true, false);
            Unlock();
        }
    }
    
    public MessageTranslator TranslatorIn {
        get {
            return iTranslatorIn;
        }
    }
    
    public MessageTranslator TranslatorOut {
        get {
            return iTranslatorOut;
        }
    }
    
    public void AddChild(Node aChild) {
        if(aChild.Parent == this) {
            throw new DuplicateNode(aChild.Fullname);
        }
        
        if(aChild.Parent != null) {
            aChild.Parent.RemoveChild(aChild);
        }
        
        aChild.Parent = this;
        Lock();
        iChildrenList.Add(new ReferencePlugin<Node>(aChild));
        //aChild.WorldTranslation = aChild.WorldTranslation;
        aChild.Update(true, true);
        Unlock();
        ObserverUpdate();
    }
    
    protected void AddCompositeChild(Node aChild) {
        if(aChild.Parent == this) {
            throw new DuplicateNode(aChild.Fullname);
        }
        
        if(aChild.Parent != null) {
            aChild.Parent.RemoveChild(aChild);
        }
        
        aChild.Parent = this;
        Lock();
        iCompositeChildrenList.Add(new ReferencePlugin<Node>(aChild));
        //aChild.WorldTranslation = aChild.WorldTranslation;
        aChild.Update(true, true);
        Unlock();
        ObserverUpdate();
    }
    
    public void RemoveChild(Node aChild) {
        if(aChild.Parent != this) {
            throw new InvalidParent();
        }
        Lock();
        ReferencePlugin<Node> obj = null;
        foreach(ReferencePlugin<Node> node in iChildrenList) {
            if(node.Object == aChild) {
                obj = node;
                break;
            }
        }
        iChildrenList.Remove(obj);
        aChild.Parent = null;
        Unlock();
        ObserverUpdate();
    }
    
    protected void RemoveCompositeChild(Node aChild) {
        if(aChild.Parent != this) {
            throw new InvalidParent();
        }
        Lock();
        ReferencePlugin<Node> obj = null;
        foreach(ReferencePlugin<Node> node in iCompositeChildrenList) {
            if(node.Object == aChild) {
                obj = node;
                break;
            }
        }
        iCompositeChildrenList.Remove(obj);
        aChild.Parent = null;
        Unlock();
        ObserverUpdate();
    }
    
    protected void ReplaceCompositeChild(uint aIndex, Node aChild) {
        if(aChild.Parent == this) {
            throw new DuplicateNode(aChild.Fullname);
        }
        if(aChild.Parent != null) {
            aChild.Parent.RemoveChild(aChild);
        }
        aChild.Parent = this;
        Lock();
        iCompositeChildrenList[(int)aIndex].Object.Parent = null;
        iCompositeChildrenList[(int)aIndex] = new ReferencePlugin<Node>(aChild);
        Unlock();
        ObserverUpdate();
    }
    
    public List<ReferencePlugin<Node>> Children {
        get {
            return iChildrenList;
        }
    }
    
    public List<ReferencePlugin<Node>> CompositeChildren {
        get {
            return iCompositeChildrenList;
        }
    }
    
    public Node Child(int aIndex) {
        Lock();
        ReferencePlugin<Node> result = iChildrenList[aIndex];
        Unlock();
        return result.Object;
    }
    
    public Node CompositeChild(int aIndex) {
        Lock();
        ReferencePlugin<Node> result = iCompositeChildrenList[aIndex];
        Unlock();
        return result.Object;
    }
    
    public virtual void Update(bool aRecurse, bool aForce) {
        bool update = false;
        if(iActive == true || aForce == true) {
            if(Parent != null) {
                Vector3d parentPos = Parent.WorldSrt.Translation;
                iWorldSrt.Translation = parentPos + iLocalSrt.Translation;
            } else {
                iWorldSrt.Translation = iLocalSrt.Translation;
            }
            //Trace.WriteLine(Trace.kGui, "Updated " + Fullname + ": LocalTrans = " + iLocalSrt.Translation + ", WorldTrans = " + iWorldSrt.Translation);
            if(aRecurse == true) {
                ReferencePlugin<Node>[] list;
                Lock();
                list = iChildrenList.ToArray();
                Unlock();
                foreach(ReferencePlugin<Node> node in list) {
                    node.Object.Update(true, aForce);
                }
                Lock();
                list = iCompositeChildrenList.ToArray();
                Unlock();
                foreach(ReferencePlugin<Node> node in list) {
                    node.Object.Update(true, aForce);
                }
            }
            update = true;
        }
        if(update) {
            ObserverUpdate();
        }
    }
    
    public virtual void Render() {
    }
    
    public virtual void Render(bool aForce) {
    }
    
    public virtual void BeginDraw() {
    }
    
    public virtual void Draw(IRenderer aRenderer) {
    }
    
    public virtual void EndDraw() {
    }
    
    public virtual void Visit(Visitor aVisitor, bool aVisitNonActive) {
        aVisitor.AcceptNode(this);
        ReferencePlugin<Node>[] list;
        Lock();
        list = iChildrenList.ToArray();
        Unlock();
        foreach(ReferencePlugin<Node> node in list) {
            if(node.Object.Active || aVisitNonActive) {
                node.Object.Visit(aVisitor, aVisitNonActive);
            }
        }
        Lock();
        list = iCompositeChildrenList.ToArray();
        Unlock();
        foreach(ReferencePlugin<Node> node in list) {
            if(node.Object.Active || aVisitNonActive) {
                node.Object.Visit(aVisitor, aVisitNonActive);
            }
        }
    }
    
    public override List<string> PackageDependencies() {
        Lock();
        List<string> deps = base.PackageDependencies();
        foreach(ReferencePlugin<Node> n in iChildrenList) {
            Trace.WriteLine(Trace.kGui, "Node.PackageDependencies: " + n.Object.Namespace);
            deps.Add(n.Object.Namespace);
        }
        foreach(Translator t in iTranslatorIn.Translators) {
            Trace.WriteLine(Trace.kGui, "Node.PackageDependencies: " + t.ToMessage.Namespace);
            deps.Add(t.ToMessage.Namespace);
            Trace.WriteLine(Trace.kGui, "Node.PackageDependencies: " + t.FromMessage.Namespace);
            deps.Add(t.FromMessage.Namespace);
        }
        foreach(Translator t in iTranslatorOut.Translators) {
            Trace.WriteLine(Trace.kGui, "Node.PackageDependencies: " + t.ToMessage.Namespace);
            deps.Add(t.ToMessage.Namespace);
            Trace.WriteLine(Trace.kGui, "Node.PackageDependencies: " + t.FromMessage.Namespace);
            deps.Add(t.FromMessage.Namespace);
        }
        Unlock();
        return deps;
    }
    
    public override void Vector3d(Vector3d aVector) {
        if(iActive) {
            base.Vector3d(aVector);
            ReferencePlugin<Node>[] list;
            Lock();
            list = iChildrenList.ToArray();
            Unlock();
            foreach(ReferencePlugin<Node> node in list) {
                node.Object.Vector3d(aVector);
            }
            Lock();
            list = iCompositeChildrenList.ToArray();
            Unlock();
            foreach(ReferencePlugin<Node> node in list) {
                node.Object.Vector3d(aVector);
            }
        }
    }
    
//  public override void Matrix(Matrix aMatrix) {
//  }

    public override bool ProcessMessage(Message aMessage) {
        if(aMessage.Fullname == Fullname) {
            MsgSetActive setActive = aMessage as MsgSetActive;
            if(setActive != null) {
                Active = setActive.Active;
                Render(false);
                return true;
            }
        }
        return false;
    }

    public virtual void Receive(Message aMessage) {
        if(ProcessMessage(aMessage)) {      // process message as it might be an active message
            Lock();
            iTranslatorIn.SendMessage(aMessage);
            Unlock();
            return;
        }
        if(Active) {        // If the node is active pass the message on to its plugin chain
            if(base.ProcessMessage(aMessage)) {
                Lock();
                iTranslatorIn.SendMessage(aMessage);
                Unlock();
                return;
            }
        }
        ReferencePlugin<Node>[] list;
        Lock();
        list = iChildrenList.ToArray();
        Unlock();
        foreach(ReferencePlugin<Node> node in list) {
            node.Object.Receive(aMessage);
        }
        Lock();
        list = iCompositeChildrenList.ToArray();
        Unlock();
        foreach(ReferencePlugin<Node> node in list) {
            node.Object.Receive(aMessage);
        }
    }
    
    public override void SendMessage(Message aMessage) {
		//Console.WriteLine(">SendMessage");
        Lock();
		//Console.WriteLine(">>SendMessage");
        iTranslatorOut.SendMessage(aMessage);
        Unlock();
		//Console.WriteLine("<<SendMessage");
        Messenger.Instance.ApplicationMessage(aMessage);
		//Console.WriteLine("<SendMessage");
    }
    
    public void ResetRootHint() {
        iRoot = false;
    }

    public Node Search(string aName)
    {
        VisitorSearch search = new VisitorSearch(aName);
        Node node = search.Search(this);
        Assert.Check(node != null);
        return (node);
    }
				
	protected void Lock() {
		//iMutex.WaitOne();
	}
	
	protected void Unlock() {
		//iMutex.ReleaseMutex();
	}
    
    private bool iRoot = false;
    protected Mutex iMutex = new Mutex(false);
    protected bool iActive = true;
    protected ReferencePlugin<Node> iParent = new ReferencePlugin<Node>();
    protected List<ReferencePlugin<Node>> iChildrenList = new List<ReferencePlugin<Node>>();
    protected List<ReferencePlugin<Node>> iCompositeChildrenList = new List<ReferencePlugin<Node>>();
    // TODO: message translators for in and out
    protected MessageTranslator iTranslatorIn = new MessageTranslator();
    protected MessageTranslator iTranslatorOut = new MessageTranslator();
    protected Srt iLocalSrt = new Srt();
    protected Srt iWorldSrt = new Srt();    
}

} // Scenegraph
} // Gui
} // Linn

File: Public/Gui/Linn/Gui/SceneGraph/Node.py
import sys
from Ft.Xml import XPath
import Linn.Gui.Serialize
import Linn.Gui.Messenger
import Linn.Gui.Renderer
import Linn.Gui.Resources.Message
import Linn.Gui.Resources.Plugin
import Linn.Gui.Resources.MessageTranslator
import Linn.Gui.SceneGraph.Srt

class DuplicateElement(Exception):
    def __init__(self, aSrcName, aName):
        self.iSrcName = aSrcName
        self.iName = aName
    def __str__(self):
        return "%s already exists in %s's list" % (self.iName, self.iSrcName)
    
class ElementNotFound(Exception):
    def __init__(self, aSrcName, aName):
        self.iSrcName = aSrcName
        self.iName = aName
    def __str__(self):
        return "%s not found in %s's list" % (self.iName, self.iSrcName)
    
class NodeNotParent(Exception):
    def __init__(self, aSrcName, aName):
        self.iSrcName = aSrcName
        self.iName = aName
    def __str__(self):
        return "%s is not %s's parent" % (self.iSrcName, self.iName)
        

class Node(Linn.Gui.Resources.Plugin.Plugin):
    def __init__(self, aName=None, aActive=True):
        if aName == None:
            aName = self.UniqueName('Node')
        Linn.Gui.Resources.Plugin.Plugin.__init__(self, aName)
        self.iParent = None
        self.iChildren = []
        self.iInMsgTrans = Linn.Gui.Resources.MessageTranslator.MessageTranslator()
        self.iOutMsgTrans = Linn.Gui.Resources.MessageTranslator.MessageTranslator()
        self.iActive = aActive
        self.iWorldSrt = Linn.Gui.SceneGraph.Srt.Srt()
        self.iLocalSrt = Linn.Gui.SceneGraph.Srt.Srt()
        self.iMsgQ = []
        
    def PackageDependencies(self):
        deps = []
        for i in self.iChildren:
            deps.append(i.Namespace())
        for i in self.iOutMsgTrans.TranslatorList():
            deps.append(i.iFromMsg.Namespace())
            deps.append(i.iToMsg.Namespace())
        return deps
        
    def Load(self, aStream):
        Linn.Gui.Resources.Plugin.Plugin.Load(self, aStream)
        
        parent = aStream.LoadNode('Parent')
        if len(parent) != 1:
            raise Linn.Gui.Serialize.InvalidLayoutFile('<Parent>', len(parent), 1)
        if parent[0].hasChildNodes():
            self.iParent = parent[0].firstChild.nodeValue
        #print 'parent =', self.iParent
        
        active = aStream.LoadNode('Active')
        if len(active) != 1:
            raise Linn.Gui.Serialize.InvalidLayoutFile('<Active>', len(active), 1)
        if active[0].hasChildNodes():
            value = active[0].firstChild.nodeValue.lower()
            if(value == '1' or value == 'true'):
                self.iActive = True
            else:
                self.iActive = False
        #print 'active =', self.iActive
        
        srt = aStream.LoadNode('Srt')
        if len(srt) != 1:
            raise Linn.Gui.Serialize.InvalidLayoutFile('<Srt>', len(srt), 1)
        aStream.PushContextNode(srt[0])
        self.iLocalSrt.Load(aStream)
        aStream.PopContextNode()
        
        msgTrans = aStream.LoadNode('OutMessageTranslator')
        if len(msgTrans) != 1:
            raise Linn.Gui.Serialize.InvalidLayoutFile('<OutMessageTranslator>', len(msgTrans), 1)
        aStream.PushContextNode(msgTrans[0])
        self.iOutMsgTrans.Load(aStream)
        aStream.PopContextNode()
                    
        children = aStream.LoadNode('Children')
        if len(children) == 1:
            children = aStream.LoadNode('Children/Child')
            for child in children:
                try:
                    self.iChildren.index(child.firstChild.nodeValue)
                    print 'Warning: Removing %s from list (duplicate)' % child.firstChild.nodeValue
                except ValueError, e:
                    self.iChildren.append(child.firstChild.nodeValue)
                #print 'child =', child.firstChild.nodeValue
        elif len(children) > 1:
            raise Linn.Gui.Serialize.InvalidLayoutFile('<Children>', len(children), 1)
    
    def Save(self, aStream):
        if aStream.ContextAttribute('class') == None:
            aStream.SetContextAttribute('class', 'Node')
            
        Linn.Gui.Resources.Plugin.Plugin.Save(self, aStream)
        
        parent = aStream.CreateElement('Parent')
        if self.iParent and self.iParent.Name() != 'SceneGraphRoot':
            parent.appendChild(aStream.CreateTextNode(self.iParent.Fullname()))
        aStream.SaveNode(parent)
        
        active = aStream.CreateElement('Active')
        active.appendChild(aStream.CreateTextNode(str(self.iActive).lower()))
        aStream.SaveNode(active)
        
        self.iLocalSrt.Save(aStream)
        
        outTranslator = aStream.CreateElement('OutMessageTranslator')
        aStream.PushContextNode(outTranslator)
        self.iOutMsgTrans.Save(aStream)
        aStream.PopContextNode()
        aStream.SaveNode(outTranslator)
        
        children = aStream.CreateElement('Children')
        if len(self.iChildren):
            for child in self.iChildren:
                childRef = aStream.CreateElement('Child')
                childRef.appendChild(aStream.CreateTextNode(child.Fullname()))
                aStream.SaveNode(childRef)
                children.appendChild(childRef)
        aStream.SaveNode(children)
        
    def Link(self):
        #print 'Linking', self.Name(), '[Node]'
        Linn.Gui.Resources.Plugin.Plugin.Link(self)
        
        if not isinstance(self.iParent, Linn.Gui.SceneGraph.Node.Node):
            self.iParent = self.LinkPluginByName(self.iParent)
        for i in range(len(self.iChildren)):
            self.iChildren[i] = self.LinkPluginByName(self.iChildren[i])
            if self.iChildren[i].Parent() == None:
                self.iChildren[i].SetParent(self.LinkPluginByName(self.Fullname()))
        
    def Parent(self):
        return self.iParent
    
    def SetParent(self, aParent):
        self.iParent = aParent
        self.ObserverUpdate()
        
    def Active(self):
        return self.iActive
    
    def SetActive(self, aActive):
        self.iActive = aActive
        for child in self.iChildren:
            child.SetActive(aActive)
        self.Update()
        self.ObserverUpdate()
        
    def WorldSrt(self):
        return self.iWorldSrt
    
    def LocalSrt(self):
        return self.iLocalSrt
    
    def SetLocalTranslation(self, aVector3d):
        self.iLocalSrt.SetTranslation(aVector3d)
        self.Update()
        self.ObserverUpdate()
        
    def SetWorldTranslation(self, aVector3d):
        self.iWorldSrt.SetTranslation(aVector3d)
        if self.iParent:
            self.iLocalSrt.SetTranslation(self.iWorldSrt.Translation() - self.iParent.LocalSrt().Translation())
        else:
            self.iLocalSrt.SetTranslation(aVector3d)
        self.Update()
        self.ObserverUpdate()
            
    def OutMsgTranslator(self):
        return self.iOutMsgTrans
    
    def AddChild(self, aChild):
        """Add child to children list.
           If the child already has a parent:
            - Remove child from current parent
            - Add child to desired parent"""
        if aChild.Parent() == self:
            self.iChildren.index(aChild)
            raise DuplicateElement(self.Fullname(), aChild.Fullname())
        
        if aChild.Parent():
            aChild.Parent().RemoveChild(aChild)
            
        # set child's parent to self and add
        aChild.SetParent(self)
        self.iChildren.append(aChild)
        aChild.Update()
        self.ObserverUpdate()
        
    def RemoveChild(self, aChild):
        """remove child from children list (if found)."""
        if aChild.Parent() != self:
            raise NodeNotParent(self.Fullname(), aChild.Fullname())
        try:
            # set child's parent to None and remove
            self.iChildren.remove(aChild)
            aChild.SetParent(None)
            self.ObserverUpdate()
        except ValueError, e:
            raise ElementNotFound(self.Fullname(), aChild.Fullname())
    
    def Child(self, aIndex):
        """obtain child at index aIndex."""
        return self.iChildren[aIndex]
    
    def NumChildren(self):
        return len(self.iChildren)
    
    def Children(self):
        return self.iChildren
    
    def Draw(self, aRenderer):
        pass
    
    def Visit(self, aVisitor):
        # accept visitor for this node
        aVisitor.AcceptNode(self)
        # visit children
        for child in self.iChildren:
            child.Visit(aVisitor)
            
    def Update(self, aRecurse=True):
        if self.iActive:
            if self.iParent:
                parentPos = self.iParent.WorldSrt().Translation()
                self.iWorldSrt.SetTranslation(parentPos + self.iLocalSrt.Translation())
            else:
                self.iWorldSrt.SetTranslation(self.iLocalSrt.Translation())
                
            if aRecurse:
                for child in self.iChildren:
                    child.Update()
            self.ObserverUpdate()
                    
    def Matrix(self, aMatrix):
        Linn.Gui.Resources.Plugin.Plugin.Matrix(self, aMatrix)
        for child in self.iChildren:
            child.Matrix(aMatrix)
        
    def Vector3d(self, aVector3d):
        Linn.Gui.Resources.Plugin.Plugin.Vector3d(self, aVector3d)
        for child in self.iChildren:
            child.Vector3d(aVector3d)
            
    def ProcessMessage(self, aMessage):
        try:
            # check to see if we have recursed with the same message
            # this occurs once a message has gone around its plugin chain
            # in this case we just dont process the message
            self.iMsgQ.index(aMessage)
            return False
        except ValueError:
            return True
            
    def ReceiveMessage(self, aMessage):
        if self.ProcessMessage(aMessage):
            # add message we are processing to the queue
            self.iMsgQ.append(aMessage)
            
            if aMessage.Fullname() == self.Fullname():
                if isinstance(aMessage, Linn.Gui.Resources.Message.MsgSetActive):
                    self.SetActive(aMessage.Active())
                self.iMsgQ.remove(aMessage)
                return
                
            Linn.Gui.Resources.Plugin.Plugin.ReceiveMessage(self, aMessage)
            for child in self.iChildren:
                child.ReceiveMessage(aMessage)
            # remove processed message from the queue
            self.iMsgQ.remove(aMessage)
                
    def SendMessage(self, aMessage):
        if not self.iOutMsgTrans.SendMessage(aMessage):
            Linn.Gui.Messenger.GetMessenger().ApplicationMessage(aMessage)


        
 
